### PR TITLE
feat pt : Support property fitting

### PR DESCRIPTION
**Pull request by Chengqian-Zhang**
Solve issue #3866

## Summary by CodeRabbit

- **New Features**
  - Introduced property fitting neural networks with the new `PropertyFittingNet` class.
  - Added the `DeepProperty` class for evaluating properties of structures using a deep learning model.
  - Implemented the `PropertyModel` class to integrate properties specific to atomic models.

- **Enhancements**
  - Added an `intensive` property to several classes to indicate whether a fitting property is intensive or extensive.
  - Enhanced output property definitions and manipulation methods in various models for improved property evaluation.
  - Expanded loss function capabilities to handle the "property" loss type.
  - Improved argument definitions for fitting and loss functionalities, enhancing configurability.
  - Updated the model selection logic to include the new `PropertyModel`.
  - Enhanced the `DeepEvalWrapper` class to support additional model evaluation features, including new methods for retrieving model characteristics.

- **Documentation**
  - Updated class docstrings to reflect new attributes and parameters, improving clarity and usability.

- **Tests**
  - Expanded the set of test examples related to the "property" category to improve test coverage.
  - Introduced new test classes and parameterized tests for improved validation of property-related functionalities.

### CODE REVIEW

**Comment by Chengqian-Zhang** [comment]
Lack `dp test` on property and doc for property fitting. You can review other codes and iterate the code together.

File: deepmd/dpmodel/fitting/property_fitting.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        return obj
+
+    def serialize(self) -> dict:
+        """Serialize the fitting to dict."""
+        dd = {
+            **InvarFitting.serialize(self),
+            "type": "property",
+            "task_dim": self.task_dim,
+        }
+        dd["@variables"]["bias_atom_e"] = self.bias_atom_e
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/fitting/property_fitting.py:88]: why do you need to store the energy bias? it should have been saved by the base class.

File: deepmd/infer/deep_property.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        )
+        property = results["property_redu"].reshape(nframes, self.get_task_dim())
+
+        if atomic:
+            return (
+                property,
+                atomic_property,
+            )
+        else:
+            return (property,)
```
Comment (wanghan-iapcm) [inline deepmd/infer/deep_property.py:132]: is the property normalized by 1/N if the system is intensive? if so where have you implemented it?

File: deepmd/pt/model/model/property_model.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            mapping,
+            fparam=fparam,
+            aparam=aparam,
+            do_atomic_virial=do_atomic_virial,
+            comm_dict=comm_dict,
+        )
+        model_predict = {}
+        model_predict["atom_property"] = model_ret["property"]
+        model_predict["property"] = model_ret["property_redu"]
+        if self.atomic_model.fitting_net.intensive:
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/model/property_model.py:95]: not good idea to directly access member variables of a class object.

File: deepmd/pt/model/atomic_model/base_atomic_model.py [hunk 4]
```diff
@@ -455,13 +455,22 @@ def change_out_bias(
             )
             self._store_out_stat(delta_bias, out_std, add=True)
         elif bias_adjust_mode == "set-by-statistic":
+            intensive = (
+                self.fitting_net.intensive
+                if (
+                    hasattr(self, "fitting_net")
+                    and hasattr(self.fitting_net, "intensive")
+                )
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/atomic_model/base_atomic_model.py:463]: not good idea to access data member of class object.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]
1. definition for the atomic model and model in the numpy backend is needed. 
2. do you really need to provide the `intensive` attribute in the fitting net? For me it should be a attribute of the model. Note that you only reduce the atomic contribution by the model, and the output should have been normalized with 1/N by the model. 
3. probably you may want to define whether a variable is intensive by the `OutputDef` class.

File: deepmd/infer/deep_property.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                [
+                    OutputVariableDef(
+                        "property",
+                        shape=[-1],
+                        reduciable=True,
+                        atomic=True,
+                    ),
+                ]
+            )
+        )
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/infer/deep_property.py:57]: @wanghan-iapcm should we replace all `reduciable` with `reducible`? The latter is the correct spelling.

File: examples/property/train/input_torch.json
Comment (njzjz) [inline examples/property/train/input_torch.json:1]: You may add a `README.md` to explain what the "property" is in this example.

File: deepmd/dpmodel/fitting/property_fitting.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            precision=precision,
+            mixed_types=mixed_types,
+            exclude_types=exclude_types,
+            type_map=type_map,
+        )
+
+    @classmethod
+    def deserialize(cls, data: dict) -> "PropertyFittingNet":
+        data = copy.deepcopy(data)
+        check_version_compatibility(data.pop("@version", 1), 2, 1)
```
Comment (njzjz) [inline deepmd/dpmodel/fitting/property_fitting.py:69]: No need to have a default value, as all models should have `@version`.

**Review by njzjz — CHANGES_REQUESTED** [review]

File: deepmd/pt/utils/stat.py [hunk 7]
```diff
@@ -240,6 +240,7 @@ def compute_output_stats(
     rcond: Optional[float] = None,
     preset_bias: Optional[Dict[str, List[Optional[torch.Tensor]]]] = None,
     model_forward: Optional[Callable[..., torch.Tensor]] = None,
+    intensive: Optional[bool] = False,
```
Comment (njzjz) [inline deepmd/pt/utils/stat.py:243]: What is the behavior of `None`? If you don't implement it, please don't add `Optional`.

**Review by njzjz — CHANGES_REQUESTED** [review]

File: deepmd/infer/deep_property.py (hunk 2 above)
Comment (Chengqian-Zhang) [reply deepmd/infer/deep_property.py:132]: The property is  normalized by 1/N if the system is intensive. It has been done in function `self.deep_eval.eval`, specifically in `PropertyModel` module. So we do not need to normalize it in `DeepProperty`.

File: deepmd/dpmodel/fitting/property_fitting.py (hunk 1 above)
Comment (Chengqian-Zhang) [reply deepmd/dpmodel/fitting/property_fitting.py:88]: I have deleted the sentence.

File: deepmd/pt/model/model/property_model.py (hunk 3 above)
Comment (Chengqian-Zhang) [reply deepmd/pt/model/model/property_model.py:95]: I have change this sentence to `if self.get_intensive()` and implemented `get_intensive()` method in `PropertyModel` class. Is it elegant？

File: deepmd/pt/model/atomic_model/base_atomic_model.py (hunk 4 above)
Comment (Chengqian-Zhang) [reply deepmd/pt/model/atomic_model/base_atomic_model.py:463]: I have implemented `get_intensive` method in `atomic_model` layer. And change this sentence to `intensive = self.get_intensive()`. Is it elegant?

File: deepmd/dpmodel/fitting/property_fitting.py (hunk 6 above)
Comment (Chengqian-Zhang) [reply deepmd/dpmodel/fitting/property_fitting.py:69]: I have changed `check_version_compatibility(data.pop("@version", 1), 2, 1)` to `check_version_compatibility(data.pop("@version"), 2, 1)`

File: deepmd/pt/utils/stat.py (hunk 7 above)
Comment (Chengqian-Zhang) [reply deepmd/pt/utils/stat.py:243]: I have deleted `Optional`

File: deepmd/infer/deep_property.py (hunk 2 above)
Comment (wanghan-iapcm) [reply deepmd/infer/deep_property.py:137]: It should be handled by the model, just like all the reducible outputs are handled by the model.

File: deepmd/dpmodel/atomic_model/property_atomic_model.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    DPAtomicModel,
+)
+
+
+class DPPropertyAtomicModel(DPAtomicModel):
+    def __init__(self, descriptor, fitting, type_map, **kwargs):
+        assert isinstance(fitting, PropertyFittingNet)
+        super().__init__(descriptor, fitting, type_map, **kwargs)
+
+    def get_intensive(self) -> bool:
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/atomic_model/property_atomic_model.py:16]: you do not need it any more if you have implement the attribute in the output def

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

**Comment by Chengqian-Zhang** [comment]
1. I add attribute `intensive` to `OutputVariableDef`. An `intensive` variable must be `reducible`. If a variable is `intensive`, doing reduce is to average the output of all the atoms instead of summing them, refer to ” deepmd/pt/model/model/transform_output.py”.
2. When calculating output bias, for `intensive` variables, we need to multiply the value of the variable by the number of atoms before performing linear regression calculation, see “deepmd/pt/utils/stat.py” for details.

**Review by njzjz — CHANGES_REQUESTED** [review]
Please add the new class to `source/tests/universal`.

File: deepmd/dpmodel/output_def.py [hunk 9]
```diff
@@ -211,13 +214,17 @@ def __init__(
         self.reducible = reducible
         self.r_differentiable = r_differentiable
         self.c_differentiable = c_differentiable
+        self.intensive = intensive
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/output_def.py:217]: the new feature is not covered by the UT.

File: deepmd/dpmodel/fitting/property_fitting.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+from deepmd.dpmodel.fitting.invar_fitting import (
+    InvarFitting,
+)
+from deepmd.utils.version import (
+    check_version_compatibility,
+)
+
+
+@InvarFitting.register("property")
+class PropertyFittingNet(InvarFitting):
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/fitting/property_fitting.py:23]: doc str for the arguments of the class init is missing.

File: deepmd/pt/model/task/property.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+)
+
+dtype = env.GLOBAL_PT_FLOAT_PRECISION
+device = env.DEVICE
+
+log = logging.getLogger(__name__)
+
+
+@Fitting.register("property")
+class PropertyFittingNet(InvarFitting):
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/property.py:38]: doc str for the arguments of the class init is missing.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

**Comment by Chengqian-Zhang** [comment]
1. Add new class to `source/tests/universal`
2. Add UT for `intensive` in `output_def`
3. Add doc str for the arguments of the `PropertyFittingNet` class init.

**Comment by Chengqian-Zhang** [comment]
I find that we lack UT for `deepmd/pt/model/model/transform_output.py` and `deepmd/dpmodel/model/transform_output.py`, I will add UT tomorrow.

File: deepmd/pt/loss/property.py
Comment (iProzd) [inline deepmd/pt/loss/property.py:1]: What about `atom_property` in loss? It seems that `atom_property` only showups in `dp test` but not in data and loss.

File: deepmd/pt/model/model/__init__.py [hunk 12]
```diff
@@ -182,6 +185,8 @@ def get_standard_model(model_params):
         modelcls = DOSModel
     elif fitting_net["type"] in ["ener", "direct_force_ener"]:
         modelcls = EnergyModel
+    elif fitting_net["type"] in ["prop", "property"]:
```
Comment (iProzd) [inline deepmd/pt/model/model/__init__.py:188]: Why two keys here?

File: deepmd/utils/argcheck.py [hunk 13]
```diff
@@ -2280,6 +2321,36 @@ def loss_dos():
     ]
 
 
+@loss_args_plugin.register("property")
+def loss_property():
+    doc_loss_func = "The loss function to minimize, such as 'mae','smooth_mae'."
+    doc_metric = "The metric such as 'mae','smooth_mae' for display."
```
Comment (iProzd) [inline deepmd/utils/argcheck.py:2327]: Please list all the options here.

File: examples/property/train/input_torch.json [hunk 14]
```diff
@@ -0,0 +1,213 @@
+{
+  "_comment": "that's all",
+  "model": {
+    "type_map": [
```
Comment (iProzd) [inline examples/property/train/input_torch.json:4]: Can we optimize the type_map here, as an example for data only in QM9?

File: source/tests/universal/dpmodel/atomc_model/test_atomic_model.py [hunk 15]
```diff
@@ -398,3 +402,67 @@ def setUpClass(cls):
...
         cls.expected_dim_aparam = ft.get_dim_aparam()
+
+
+@parameterized(
+    des_parameterized=(
+        (
+            *[(param_func, DescrptDPA1) for param_func in DescriptorParamDPA1List],
+            *[(param_func, DescrptDPA2) for param_func in DescriptorParamDPA2List],
+            (DescriptorParamHybridMixed, DescrptHybrid),
+            (DescriptorParamHybridMixedTTebd, DescrptHybrid),
```
Comment (iProzd) [inline source/tests/universal/dpmodel/atomc_model/test_atomic_model.py:413]: Why there's test for se_a models in pt but not here in dp?

**Review by iProzd — CHANGES_REQUESTED** [review]
UTs for dp test is needed.

**Review by wanghan-iapcm — APPROVED** [review]

File: deepmd/pt/model/model/__init__.py (hunk 12 above)
Comment (Chengqian-Zhang) [reply deepmd/pt/model/model/__init__.py:188]: I think that it can prevent users from writing ‘prop’ by mistake.  I will remove the ‘prop’ option.

File: deepmd/utils/argcheck.py (hunk 13 above)
Comment (Chengqian-Zhang) [reply deepmd/utils/argcheck.py:2327]: I have changed it to:
"The metric for display. This list can include 'smooth_mae', 'mae', 'mse' and 'rmse'."

File: examples/property/train/input_torch.json [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+      "Cn",
+      "Nh",
+      "Fl",
+      "Mc",
+      "Lv",
+      "Ts",
+      "Og"
+    ],
+    "descriptor": {
+      "type": "dpa2",
```
[the remark this answers is not in the document]
Comment (Chengqian-Zhang) [reply examples/property/train/input_torch.json]: done

File: source/tests/universal/dpmodel/atomc_model/test_atomic_model.py (hunk 15 above)
Comment (Chengqian-Zhang) [reply source/tests/universal/dpmodel/atomc_model/test_atomic_model.py:414]: I add it

File: deepmd/pt/loss/property.py
Comment (Chengqian-Zhang) [reply deepmd/pt/loss/property.py:1]: Util now, `atom_property` is almost useless, it is only used to compute `property`.

Comment (Chengqian-Zhang) [reply deepmd/pt/loss/property.py:1]: What's the best way to handle `atom_property`?

**Comment by njzjz** [comment]
> ```diff
> + Hits        42870    43087     +217     
> - Misses       7822     7934     +112     
> - Partials      946      948       +2     
> ```

I see that 1/3 of the new codes have not been tested.

**Comment by Chengqian-Zhang** [comment]
> > ```diff
> > + Hits        42870    43087     +217     
> > - Misses       7822     7934     +112     
> > - Partials      946      948       +2     
> > ```
> 
> I see that 1/3 of the new codes have not been tested.

Where can I see these datas about missing UT? I will add UT based on these datas.

**Review by iProzd — APPROVED** [review]

**Comment by njzjz** [comment]
> > > ```diff
> > > + Hits        42870    43087     +217     
> > > - Misses       7822     7934     +112     
> > > - Partials      946      948       +2     
> > > ```
> > 
> > 
> > I see that 1/3 of the new codes have not been tested.
> 
> Where can I see these datas about missing UT? I will add UT based on these datas.

You can click the link sent by @codecov or click the codecov checks.

It looks like the loss module is not tested. @iProzd Do we have a universal test fixture for loss functions?

**Comment by iProzd** [comment]
> > > > ```diff
> > > > + Hits        42870    43087     +217     
> > > > - Misses       7822     7934     +112     
> > > > - Partials      946      948       +2     
> > > > ```
> > > 
> > > 
> > > I see that 1/3 of the new codes have not been tested.
> > 
> > 
> > Where can I see these datas about missing UT? I will add UT based on these datas.
> 
> You can click the link sent by @codecov or click the codecov checks.
> 
> It looks like the loss module is not tested. @iProzd Do we have a universal test fixture for loss functions?

Not yet, maybe we need a discussion to design a universal test for loss modules.

**Review by njzjz — APPROVED** [review]
I create #4105. We may resolve the issue later instead of doing in this issue.
